### PR TITLE
WIP: Port parser.h

### PR DESCRIPTION
**Pull request by krobelus**

### CODE REVIEW

**Comment by ridiculousfish** [comment]
I have `env.cpp` [pretty far along](https://github.com/ridiculousfish/fish-shell/tree/riir-env) though I'm happy to drop it if your implementation is better!

**Comment by krobelus** [comment]
This is what's missing, we can port it in roughly this order.
We can split up some of the larger groups (but that means we'd need to define extra FFI bridges).

- [ ] src/wcstringutil.h

- [ ] src/path.h

- [ ] src/parse_util.h

- env
  - [ ] src/env.h
  - [ ] src/env_universal_common.h
  - [ ] src/env_dispatch.h
  - [ ] src/null_terminated_array.h
  - [ ] src/utf8.h (trivial to replace with Rust stdlib)

- [ ] src/autoload.h

- [ ] src/parser_keywords.h

- parse/execution
  - [ ] src/complete.h
  - [ ] src/exec.h
  - [ ] src/expand.h
  - [ ] src/function.h
  - [x] src/operation_context.h
  - [ ] src/parse_execution.h
  - [ ] src/parser.h
  - [ ] src/postfork.h
  - [x] src/proc.h
  - [ ] src/wildcard.h

- [ ] src/highlight.h

- [ ] src/history_file.h
- [ ] src/history.h

- [ ] src/fish_indent_common.h

- [ ] src/kill.h

- [ ] src/print_help.h

- reader
  - [ ] src/input.h
  - [ ] src/input_common.h
  - [ ] src/output.h
  - [ ] src/pager.h
  - [ ] src/reader.h
  - [ ] src/reader.h
  - [ ] src/screen.h

- [ ] src/builtins/

File: fish-rust/src/parser_keywords.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+pub fn parser_keywords_is_subcommand(cmd: &wstr) -> bool {
+    SUBCOMMAND_KEYWORDS.contains(&cmd) || SKIP_KEYWORDS.contains(&cmd)
+}
+
+/// Tests if the specified command is a reserved word, i.e. if it is the name of one of the builtin
+/// functions that change the block or command scope, like 'for', 'end' or 'command' or 'exec'.
+/// These functions may not be overloaded, so their names are reserved.
+///
+/// \param word The command name to test
+/// \return 1 of the command parameter is a command, 0 otherwise
```
Comment (faho) [inline fish-rust/src/parser_keywords.rs:54]: (this returns true and false, not 1 and 0)

Comment (krobelus) [reply fish-rust/src/parser_keywords.rs:54]: yeah I let some docstrings go stale. These two lines we can remove altogether

**Comment by krobelus** [comment]
moved to the `riir-parser` branch

**Pull request closed without merging**